### ISSUE TITLE
Inconsistent register state during HOOK_BLOCK callback

### ISSUE THREAD

**Issue by rhelmot** [unicorn-engine/unicorn#1643]
I'm running into an issue where the cpu register state (particularly the instruction pointer) is not correct during a block callback:

```
#0  hook_block (uc=0x23c3de0, address=134513775, size=10, user_data=0x26332e0) at sim_unicorn.cpp:2614
2614            State *state = (State *)user_data;
(rr) p/x address
$95 = 0x804846f
(rr) p/x ((X86CPU *)uc->cpu)->env->eip
$96 = 0x8048473
```

(the correct answer is reported as the address argument to the callback)

Is this expected? angr's unicorn support has an absolutely critical dependency on being able to snapshot state with uc_context_save during block callbacks, and if the register file is inconsistent this breaks.

**Comment by rhelmot** [comment]
I can give you a testcase which reproduces this, but it will involve running angr. Is this acceptable?

**Comment by wtdcode** [comment]
Hmmm, this is expected as the PC(`env->eip`) is not sync-ed for `UC_HOOK_BLOCK`. Does it cause any break for angr?

**Comment by rhelmot** [comment]
Yes, it causes angr to break. Would you accept a patch that syncs it for UC_HOOK_BLOCK?

**Comment by wtdcode** [comment]
> Yes, it causes angr to break. Would you accept a patch that syncs it for UC_HOOK_BLOCK?

So the context saving within `UC_HOOK_BLOCK` has a wrong PC after restoring, correct? I might try to reproduce it for a unit test.

The sync is not that easy to implement as the `UC_HOOK_BLOCK` is added in the translator and not aware of the actual architecture. Thus it's hard to generate the corresponding tcg code to sync it runtime if we don't move `UC_HOOK_BLOCK` generate to `qemu/target/<arch>/translate.c`.

**Comment by rhelmot** [comment]
> the context saved within UC_HOOK_BLOCK has a wrong PC after restoring, correct?

This is correct.

> The sync is not that easy to implement

Pushing the HOOK_BLOCK codegen into the architecture's lifter seems suboptimal. It might be better to copy or factor out code from elsewhere that sets the current instruction pointer - I think I saw uc_emu_stop or some dependency thereof doing this?

**Comment by rhelmot** [comment]
Still an issue.

**Comment by rhelmot** [comment]
Still an issue.

**Comment by liangjs** [comment]
Is this really an expected behavior? The [document](https://github.com/unicorn-engine/unicorn/wiki/FAQ#why-do-i-get-a-wrong-pc-after-emulation-stops) says "the PC is sync-ed at the basic block boundary".

**Comment by rhelmot** [comment]
Still an issue. I would like to point out that it's pretty insulting to have to defend this fact to a bot every 60 days.

**Comment by wtdcode** [comment]
> Still an issue. I would like to point out that it's pretty insulting to have to defend this fact to a bot every 60 days.

My bad, I forget to add bug labels to exclude this issue from bots' check. Anyway, sorry for the delay because I'm lacking time since the start of this year.

Regarding the issue itself, I have a fix actually but it's not published as it only works for x86. I will get it work for all architectures.

**Comment by wtdcode** [comment]
Fixed in [0c512f9](https://github.com/unicorn-engine/unicorn/commit/0c512f91a176125426cb0441fc6ea316ea50dfa2) and will be included in 2.1.2. I will also add a few tests for other minor architectures.

Sorry for the long-time delay.
